components/script/microdata.rs

```diff
@@ -0,0 +1,161 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::codegen::Bindings::DocumentBinding::DocumentBinding::DocumentMethods;
+use dom::bindings::codegen::Bindings::ElementBinding::ElementBinding::ElementMethods;
+use dom::bindings::inheritance::Castable;
+use dom::bindings::root::DomRoot;
+use dom::characterdata::CharacterData;
+use dom::document::Document;
+use dom::element::Element;
+use dom::htmlelement::HTMLElement;
+use dom::node::Node;
+use dom::text::Text;
+use serde_json;
+use std::borrow::Cow;
+use std::collections::HashMap;
+
+pub struct Microdata {}
+
+impl Microdata {
+    pub fn parse(doc: &Document, node: &Node) -> HashMap<String, String> {
+        let serialized_vcard = Self::parse_vcard(doc);
+        let serialized_json = Self::parse_json(node);
+        let mut serialized_data: HashMap<String, String> = HashMap::new();
+        serialized_data.insert("vcard".to_string(), serialized_vcard);
+        serialized_data.insert("json".to_string(), serialized_json);
+        return serialized_data;
+    }
+
+    pub fn parse_vcard(doc: &Document) -> String {
+        let ele = doc.upcast::<Node>();
+        let mut start_vcard = false;
+        let mut result: String = String::new();
+        let mut master_map: HashMap<String, HashMap<String, String>> = HashMap::new();
+        let mut master_key: String = String::new();
+
+        result += "BEGIN:VCARD\nPROFILE:VCARD\nVERSION:4.0\nSOURCE:";
+        result += doc.url().as_str();
+
+        let title = doc.Title();
+        if !title.is_empty() && !title.trim().is_empty() {
+            result += "\nNAME:";
+            result += title.trim();
+        }
+
+        result += "\n";
+
+        for element in ele.traverse_preorder().filter_map(DomRoot::downcast::<Element>) {
+            if element.is::<HTMLElement>() {
+                if element.has_attribute(&local_name!("itemtype")) {
+                    let mut atoms = element.get_tokenlist_attribute(&local_name!("itemtype"), );
+                    if !atoms.is_empty() {
+                        let val = atoms.remove(0);
+                        if val.trim() == "http://microformats.org/profile/hcard" {
+                            if !start_vcard {
+                                start_vcard = true;
+                            } else {
+                                break;
+                            }
+                        }
+                    }
+                }
+                if start_vcard {
```

Comment:
The implementation here doesn't appear to follow the algorithms defined in the specification. I would expect to see code that follows a pattern like this:
```rust
// https://html.spec.whatwg.org/multipage/#extracting-a-vcard
// Steps 1 & 2
let elements = ele.traverse_preorder().filter_map(DomRoot::downcast::<HTMLElement>);
let node = elements.find(|element| {
    element.Itemtypes().iter().find(|val| val == "http://microformats.org/profile/hcard").is_some())
});
let node = match node {
    Some(node) => node,
    None => /* return no vcard */,
};
/* prepare the vcard output */
// Step 11
for element in collect_properties_of_item(node) {
    for property in element.PropertyNames() {
        // Step 11.2
        let value = property_value(element, property);
        /* remaining steps */
    }
}
/* finish up vcard output */
```
Does that make sense?